tools/compiler/src/processor.rs

```diff
@@ -60,21 +60,53 @@ pub async fn generate_registry() -> AppResult<()> {
     Ok(())
 }
 
+/// Given `ibc_data` describing a channel and `source_asset` on the source chain,
+/// compute the metadata for the asset when it is transported along the channel onto a Penumbra chain.
+fn transport_metadata_along_channel(ibc_data: &IbcConfig, source_asset: Metadata) -> Metadata {
+    // The `Metadata` structure doesn't allow modifying the internals, so drop to raw proto data
+    let mut pb_metadata: pb::Metadata = source_asset.into();
+    tracing::debug!(?pb_metadata, "original");
+
+    let prefix_channel = |x: &mut String| {
+        *x = format!("transfer/{}/{}", ibc_data.ibc_channel, x);
+    };
+
+    // Prefix the channel to the base denom.
+    prefix_channel(&mut pb_metadata.base);
+    // Prefix the channel to the display denom.
+    prefix_channel(&mut pb_metadata.display);
+    // Prefix the channel to all denom units.
+    for denom_unit in pb_metadata.denom_units.iter_mut() {
+        prefix_channel(&mut denom_unit.denom);
+    }
+
+    // Delete the asset ID, so that it will be recomputed with the adjusted base denom.
+    // Without this, decoding will fail because the asset ID won't match.
+    pb_metadata.penumbra_asset_id = None;
+
+    tracing::debug!(?pb_metadata, "new");
+    Metadata::try_from(pb_metadata).unwrap()
```

Comment:
Think we should return a result here. I'll get this in a later PR though.

Reply:
Don't think so, if we had valid Metadata to start, our modifications should also be valid, or else the compiler should fail.